lib/checkother.cpp

```diff
@@ -510,8 +510,18 @@ void CheckOther::checkRedundantAssignment()
                 else
                     start = tok->findExpressionStartEndTokens().second->next();
 
+                const Token * tokenToCheck = tok->astOperand1();
+
+                // Check if we are working with union
+                const Token * tempToken = tokenToCheck;
```

Comment:
if you make a for loop here the tempToken scope can be limited..

    for (const Token* tempToken = tokenToCheck; Token::Match(tempToken, ".");) {
        ...
    }

A token can never be "->" in cppcheck token list so you just need to check for ".".